src/System.IO/tests/BinaryWriter/BinaryWriterTests.cs

```diff
@@ -0,0 +1,345 @@
+            using (MemoryStream mstr = new MemoryStream())
+            using (BinaryWriter dw2 = new BinaryWriter(mstr))
+            using (BinaryReader dr2 = new BinaryReader(mstr))
+            {
+                dw2.Write(true);
+                dw2.Flush();
+                mstr.Position = 0;
+
+                Assert.True(dr2.ReadBoolean());
+            }
+        }
+
+        [Fact]
+        public static void BinaryWriter_CtorAndWriteTests1_Negative()
+        {
+            // [] Should throw ArgumentNullException for null argument
+            Assert.Throws<ArgumentNullException>(() => { BinaryWriter writer = new BinaryWriter(null); });
+
+            // [] Can't construct a BinaryWriter on a readonly stream
+            using (MemoryStream memStream = new MemoryStream(new byte[10], false))
+            {
+                Assert.Throws<ArgumentException>(() => { BinaryWriter writer = new BinaryWriter(memStream); });
+            }
+
+            // [] Can't construct a BinaryWriter with a closed stream
+            {
+                MemoryStream memStream = new MemoryStream();
+                memStream.Dispose();
+                Assert.Throws<ArgumentException>(() => { BinaryWriter dw2 = new BinaryWriter(memStream); });
+            }
+        }
+
+        [Theory]
+        [MemberData("EncodingAndEncodingStrings")]
+        public static void BinaryWriter_EncodingCtorAndWriteTests(Encoding encoding, string testString)
+        {
+            using (MemoryStream memStream = new MemoryStream())
+            using (BinaryWriter writer = new BinaryWriter(memStream, encoding))
+            using (BinaryReader reader = new BinaryReader(memStream, encoding))
+            {
+                writer.Write(testString);
+                writer.Flush();
+                memStream.Position = 0;
+
+                Assert.Equal(testString, reader.ReadString());
+            }
+        }
+
+        public static IEnumerable<object[]> EncodingAndEncodingStrings
+        {
+            get
+            {
+                yield return new object[] { Encoding.UTF8, "This is UTF8\u00FF" };
+                yield return new object[] { Encoding.BigEndianUnicode, "This is BigEndianUnicode\u00FF" };
+                yield return new object[] { Encoding.Unicode, "This is Unicode\u00FF" };
+            }
+        }
+
+        [Fact]
+        public static void BinaryWriter_EncodingCtorAndWriteTests_Negative()
+        {
+            // [] Check for ArgumentNullException on null stream
+            Assert.Throws<ArgumentNullException>(
+                () => { new BinaryReader((Stream)null, Encoding.UTF8); });
+
+            // [] Check for ArgumentNullException on null encoding
+            Assert.Throws<ArgumentNullException>(
+               () => { new BinaryReader(new MemoryStream(), null); });
+        }
+
+        [Fact]
+        public static void BinaryWriter_SeekTests()
```

Comment:
Per other comments, I'm going to leave this as it was in it's general form from the initial code being ported by Matt.  There are many such things that can be cleaned up subsequently.